Fix all linter warnings and errors in test suite

DerGuteMoritz opened the pull request:
Decided to start small since I already encountered a few issues that I wanted to coordinate on before continuing. I will add comments on the diff to discuss these!

DerGuteMoritz commented on .clj-kondo/config.edn:
```diff
@@ -1 +1,3 @@
-{:config-paths ["../resources/clj-kondo.exports/clj-commons/aleph"]}
+{:config-paths ["../resources/clj-kondo.exports/clj-commons/aleph"]
+ :linters      {:unsorted-required-namespaces {:level :warning}}
```
This linter isn't enabled by default but I think it's a nice one to have. And since we're likely going to have to touch pretty much all `ns` forms anyway for cleanup ... but feel free to disagree, of course!

KingMob replied:
Agreed. I prefer sorted namespaces, and they're especially cluttered in aleph/manifold.

Sorting should be safe, but in general, my one major concern is that we never take linter warnings as gospel, because there's some really funky stuff in there. Like, namespaces might be required for `extend-type` side effects, but never refer to any vars. And in one case I know of, Zach temporarily switched namespaces in the middle of a file, which plays havoc with modern editors/linters.

DerGuteMoritz replied:
> Agreed. I prefer sorted namespaces, and they're especially cluttered in aleph/manifold.

Alright, thanks for the feedback!

> Sorting should be safe, but in general, my one major concern is that we never take linter warnings as gospel, because there's some really funky stuff in there.

Sure thing! I will judiciously apply `#_:clj-kondo/ignore` comments :smile: 

> Like, namespaces might be required for extend-type side effects, but never refer to any vars.

`clj-kondo` has a convention for dealing with this - only gotta be careful to apply it where needed :smile: Thanks for the headsup!

DerGuteMoritz commented on .clj-kondo/config.edn:
```diff
@@ -1 +1,3 @@
-{:config-paths ["../resources/clj-kondo.exports/clj-commons/aleph"]}
+{:config-paths ["../resources/clj-kondo.exports/clj-commons/aleph"]
+ :linters      {:unsorted-required-namespaces {:level :warning}}
+ :lint-as      {manifold.deferred/let-flow clojure.core/let}}
```
Ideally this would be provided by `manifold`, of course. There are a few more like this coming up. Right now I tend to collect all of them here and once everything is green, extract them out and file PRs against the other projects. Sound good?

KingMob replied:
That's fine.

`let-flow'` should be linted the same way.

DerGuteMoritz commented on test/aleph/http/core_test.clj:
```diff
@@ -1,11 +1,9 @@
   (:require
-    [aleph.http.core :as core])
+   [aleph.http.core :as core]
+   [clojure.test :refer [deftest is]])
```
As you can see, I changed the indentation of these forms. That's because I used https://github.com/clojure-emacs/refactor-nrepl's `clean-ns` refactoring which produces this indentation by default. I did a quick research on how to change this but couldn't come up with anything so far. Let me know if you mind, otherwise I'll make it work *somehow* :smile:

DerGuteMoritz replied:
FTR: I just figured out how to adapt to it! Turns out that this is actually not `refactor-nrepl`'s job bur rather `clojure-mode`'s. So yeah, let me know if you'd like to preserve this!

KingMob replied:
The new 1-space indentation is more standard, according to https://guide.clojure.style/#line-break-ns-declaration.

We should probably make a single reformat-only PR. Otherwise the indentation either never gets fixed, or it gets fixed bit-by-bit and clutters up PR diffs with formatting changes.

DerGuteMoritz replied:
Alright, 1 space it is. If you don't mind, I would already apply this for the test suite in this PR. The next one will then do the same for the main code.

DerGuteMoritz commented on test/aleph/http/multipart_test.clj:
```diff
@@ -1,20 +1,18 @@
 
 (def file-to-send (File. (str (System/getProperty "user.dir") "/test/file.txt")))
 
+#_{:clj-kondo/ignore [:deprecated-var]}
```
Almost all tests in this namespace are tagged with this since they all test `mp/encode-body` rather than its apparent successor `mp/encode-request`. Maybe this is an oversight that should be corrected separately?

KingMob replied:
Hmmm. Let's leave this for now.

DerGuteMoritz commented on test/aleph/http_test.clj:
```diff
@@ -420,14 +403,14 @@
   (with-handler basic-handler
-    (let [pool (http/connection-pool {:connection-options {:response-buffer-size 16}})]
-      (dotimes [i 1 #_1e6]
+    (let [_pool (http/connection-pool {:connection-options {:response-buffer-size 16}})]
```
At first I thought to remove the whole `let` form but then I figured that perhaps it's in here as a convenience for tweaking the test somehwo?

KingMob replied:
Base on git blame, this was a partial attempt at deciphering clj-commons/aleph#193. I don't think it works as intended, so maybe the whole thing can be commented-out for now.

DerGuteMoritz replied:
Thanks for checking! I dug some more and found that this `let` was likely just forgotten to be removed with c031a64cd5110efa5585c7700e8f7de5648ac429. I'll just do that in a separate commit for clarity!

DerGuteMoritz commented on test/aleph/http_test.clj:
```diff
@@ -102,14 +90,14 @@
   {:status 200
    :body (->> (s/periodically 0.1 #(byte-array 1024))
-           (s/transform (take 1e3)))})
+              (s/transform (take (long 1e3))))})
```
Ah, right, thread macros are another instance of indentation divergence. I only changed it when I had to touch the form to fix any linter warnings, though. Note that the style I changed it to is the one recommended by [The Clojure Style Guide](https://guide.clojure.style/#threading-macros-alignment) and happens to be the default of [`clojure-mode`](https://github.com/clojure-emacs/clojure-mode). But now I know how to adjust this, so if you'd prefer, I'm happy to adapt.

KingMob replied:
Nono, I prefer the general style guide, thanks for this.

KingMob commented:
@DerGuteMoritz What's the status of this? I was looking forward to it

KingMob commented:
No worries, take your time.

On Mon, Apr 11, 2022 at 6:56 PM Moritz Heidkamp ***@***.***>
wrote:

> @KingMob <https://github.com/KingMob> I'll get back to it soon - work got
> a bit busy after I opened this but should be back in more quiet waters in a
> few days!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/clj-commons/aleph/pull/583#issuecomment-1094905607>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHHB5MCNILMLV4NF4WUQO3VEQAPLANCNFSM5REK6BCQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

KingMob commented:
@DerGuteMoritz Any updates?

DerGuteMoritz commented on test/aleph/classloader_test.clj:
```diff
@@ -39,7 +41,7 @@
   (testing "classloader: ensure the class loader is always a DynamicClassLoader"
     (let [result (CompletableFuture.)]
-      (with-dynamic-redefs [netty/operation-complete (partial operation-complete result)]
+      (with-dynamic-redefs [#_{:clj-kondo/ignore [:private-call]} netty/operation-complete (partial operation-complete result)]
```
@KingMob Had to add this one after the rebase - not sure if the var being private is intended here?

arnaudgeiser replied:
That's not something we want to expose externally, Aleph is already exposing way to many public `vars` in my opinion.
Having this `:private-call` is fine for me.

KingMob replied:
Can that be set at another level? I’d prefer not to have kondo annotations deep in code if possible.

KingMob replied:
@DerGuteMoritz Was there a way to set this elsewhere?

DerGuteMoritz replied:
Even better: Looks like adding https://github.com/clj-commons/aleph/pull/583/files#diff-d8bfac7f9aa64d7a63a2c6810095e91e716527266288a210b3f31a2c141ce3f2R3 made it completely unnecessary :smile:

DerGuteMoritz commented:
@KingMob Sorry for the delay :bow:  Back on track now. I've rebased and adjusted for the latest `master` changes and incorporated your feedback from above (see threads there for some further details). I want to look into adding a circle-ci job for `clj-kondo` and then this could be merged!

DerGuteMoritz commented:
Alright, just added a `lint` job to the Circle CI config. Don't know how to test it, though - it doesn't seem to be picked up automatically for PR checks, at least. Hope you know what to do :pray:

arnaudgeiser reviewed (approved):
Fine for me,
A follow-up for this PR would be to format using `cljfmt`.
Using any `parinfer/paredit` is really complicated on the state of the codebase.

Regarding the Github action, I suppose @KingMob will fix it afterwards if needed. :)

KingMob commented:
I need to check with borkdude on using kondo for circleci; I think Eastwood is preferred for non-interactive CI linting because it more closely mimics how the real compiler sees code. Let me get back to you on that.

KingMob commented:
> Alright, just added a `lint` job to the Circle CI config. Don't know how to test it, though - it doesn't seem to be picked up automatically for PR checks, at least. Hope you know what to do 🙏

@DerGuteMoritz I think the reason it's not picked up is the indentation level makes it a separate job, but we probably want to place it under the existing `build` job as another step (preferably right before the testing step). In general, clj-commons has used just two circleci jobs, one for building and one for deploying (deployment jobs don't yet exist for aleph and manifold for historical reasons, since their deployment isn't yet under the clj-commons maven group).

KingMob commented:
I see plenty of people using kondo for CI, so that's fine. It might be wiser to make a separate ci config file, as described in https://github.com/clj-kondo/clj-kondo/blob/master/doc/ci-integration.md

The goal should be to get it in place now, with nothing blocking merges. Aleph and Manifold do some funky stuff sometimes, and kondo should not block current development just because of that. As we modernize the codebase, we can revisit what warnings we want to turn back on.

DerGuteMoritz commented:
> A follow-up for this PR would be to format using `cljfmt`.
> Using any `parinfer/paredit` is really complicated on the state of the codebase.

Indeed, the formatting is quite idiosyncratic in places, mainly due to all these macros I think. Running `cljfmt` on it sounds like a nice follow-up indeed. I'll see if I can tackle that, too, once we've got the linting in :smile_cat: 


> @DerGuteMoritz I think the reason it's not picked up is the indentation level makes it a separate job, but we probably want to place it under the existing `build` job as another step (preferably right before the testing step).

I broke it out into a separate job for two reasons:

1. This way, the lint job will show up as a separate check on a PR, giving immediate indication of what kind of failure there is.
2. A different Docker image is required for running `clj-kondo`. AFAIK there is no way to combine two existing images into one so to achieve that one would have to either build a custom image or install `clj-kondo` from within the job which is pretty messy.

> In general, clj-commons has used just two circleci jobs, one for building and one for deploying (deployment jobs don't yet exist for aleph and manifold for historical reasons, since their deployment isn't yet under the clj-commons maven group).

And is there a technical rationale for doing this or just tradition / convention? :smile: 


> I see plenty of people using kondo for CI, so that's fine. 

Yup, the way it operates is pretty much intended for non-interactive use. Also, without an automated job like this, it's pretty much guaranteed that the code will evolve into a non-linted state pretty quickly unless reviewers diligently run the linter before merging. And since that's so trivially automated ... :smile:

> It might be wiser to make a separate ci config file, as described in https://github.com/clj-kondo/clj-kondo/blob/master/doc/ci-integration.md

This would only be needed if we want CI-specific config which currently doesn't seem necessary nor particularly desirable - why would you want to have different linting rules for development vs. CI? :thinking: 

That being said, I'm happy to adjust any of the above to your liking, just gimme the word :+1:

DerGuteMoritz commented:
Sorry, missed on bit:

> The goal should be to get it in place now, with nothing blocking merges. Aleph and Manifold do some funky stuff sometimes, and kondo should not block current development just because of that. As we modernize the codebase, we can revisit what warnings we want to turn back on.

If by that you mean that lint failures should not block PRs from getting merged, that would be another case in point for having it be a separate CI job so that it can be marked as non-blocking.

KingMob commented:
We can try adding automatic cljfmt, see how it goes. In my limited experience with it, it requires a bunch of config for good results, especially around macros.

I'm fine with treating linting as a separate job, as long as we can ensure dependencies. E.g., if either linting or building/testing fail, a deploy job should not be runnable (see other clj-commons repos for examples). Regardless, people should always be able to click into CircleCI for details, so making it directly visible in GH isn't a big priority to me.

> And is there a technical rationale for doing this or just tradition / convention?

I don't know if there's a good reason the build job can't be split up, but I do know the other clj-commons deployment jobs run on different git tags, and the branch/tag criteria is set at the job level. So...maybe?

> A different Docker image is required for running clj-kondo. 

I didn't notice you used a different docker image. Potentially dumb question: does this work correctly in circleci? I know in other clj-commons Circle setups, we download a Babashka script for deployment. We could do something similar to download kondo if necessary. (We should also upgrade to the [newer circleci clojure img](https://circleci.com/developer/images/image/cimg/clojure)...)

> If by that you mean that lint failures should not block PRs from getting merged, that would be another case in point for having it be a separate CI job so that it can be marked as non-blocking.

But some rules SHOULD be hard failures (e.g., arity errors, unresolved symbols), just not necessarily all. Here's an example of why we might want separate CI and dev configs: https://github.com/clj-commons/manifold/blob/dbdd600c5a48df314e3ad06fe2986e05e5af21c1/src/manifold/time.clj#L126-L128. Zach does some crazy ns stuff here. At the developer level, I want to be reminded to fix this, but I don't want to have this block CI, because it may take some time to fix. 

Validating correctness is more important, but as you point out, maintaining good style requires regular linting, too. We could make all the changes at once and then enable them as blockers in the same PR, but I'm afraid this is turning out to be a lot of work already. I prefer smaller, more focused PRs; they're easier to finish, easier to review, and less risky to merge. 

If you don't want to have separate dev/CI configs (which is fair), let's still try to split up this work into the following PRs: 

1. A kondo and kondo-based fix PR, no circleCI
2. A cljfmt merge that does nothing but update the indentation based on the Clojure style guide and specify the indentation rules
3. A circleCI PR that adds cljfmt and kondo jobs or steps

DerGuteMoritz commented:
> We can try adding automatic cljfmt, see how it goes. In my limited experience with it, it requires a bunch of config for good results, especially around macros.

Yeah, this will likely be some work - same with `clj-kondo` actually :grimacing:

> I'm fine with treating linting as a separate job, as long as we can ensure dependencies. E.g., if either linting or building/testing fail, a deploy job should not be runnable (see other clj-commons repos for examples).

Ah, right, looks like we would have to set up a workflow for this, similar to what's done e.g. here: https://github.com/clj-commons/byte-streams/blob/master/.circleci/config.yml#L7-L22 - looks like it should be possible to just add the lint job as another `requires` dependency to the deploy job.

> I didn't notice you used a different docker image. Potentially dumb question: does this work correctly in circleci? 

Fairly sure! [Docs say this about `image`](https://circleci.com/docs/configuration-reference#docker): "The name of a custom docker image to use. The first image listed under a job defines the job’s own primary container image where all steps will run."


> I know in other clj-commons Circle setups, we download a Babashka script for deployment. We could do something similar to download kondo if necessary.

Right, that's the approach I referred to as "messy" above :smile:  

> But some rules SHOULD be hard failures (e.g., arity errors, unresolved symbols), just not necessarily all. Here's an example of why we might want separate CI and dev configs: https://github.com/clj-commons/manifold/blob/dbdd600c5a48df314e3ad06fe2986e05e5af21c1/src/manifold/time.clj#L126-L128. Zach does some crazy ns stuff here. At the developer level, I want to be reminded to fix this, but I don't want to have this block CI, because it may take some time to fix.

OK makes sense! But I think I would rather have the default developer config be the same as the one used by CI so that one doesn't get all kinds of warnings during development which end up being irrelevant. So far I didn't have to disable any linting rules globally but once we do, we could provide a separate config which one could use instead for "full" linting when desired.


> We could make all the changes at once and then enable them as blockers in the same PR, but I'm afraid this is turning out to be a lot of work already. I prefer smaller, more focused PRs; they're easier to finish, easier to review, and less risky to merge.

Aye, agree with you here. That's why I decided to make this first PR about the test suite only so that we can settle these discussions before having invested a ton of work on the main code base :sweat_smile: 

> If you don't want to have separate dev/CI configs (which is fair), let's still try to split up this work into the following PRs:
> 
> 1. A kondo and kondo-based fix PR, no circleCI
> 2. A cljfmt merge that does nothing but update the indentation based on the Clojure style guide and specify the indentation rules
> 3. A circleCI PR that adds cljfmt and kondo jobs or steps

Sounds like a plan :+1: I will drop the CI job from this PR then. Shall we go ahead with merging it as-is, i.e. test suite only or would you prefer a big one for the whole code base now that we've mostly settled the nitty-gritty? FWIW I think I would prefer to merge this one as-is because it already provides some value and it will probably take a while for me to fix up the main code.

KingMob commented on .clj-kondo/config.edn:
```diff
@@ -1 +1,5 @@
+{:config-paths ["../resources/clj-kondo.exports/clj-commons/aleph"]
+ :linters      {:unsorted-required-namespaces {:level :warning}}
+ :lint-as      {dynamic-redef.core/with-dynamic-redefs clojure.core/with-redefs
+                manifold.deferred/let-flow clojure.core/let
+                manifold.deferred/loop clojure.core/let}}
```
Minor: this is fine for now, but there's a way to export kondo linting notes as part of a lib, so we should really add that to manifold instead. That way everything using Manifold will get the correct linting.

DerGuteMoritz replied:
Yep, already on my radar: https://github.com/clj-commons/aleph/pull/583#discussion_r830513450 :+1:

KingMob reviewed (approved):
Approved, but could you look at my last few comments before we merge?

DerGuteMoritz commented:
@KingMob There you go!
